### DISCUSSION TITLE
Differenzbildung bei Lieferungen

### DISCUSSION THREAD

**Discussion by Philippluca** [geowerkstatt/geopilot#318] [Ideas]
Am [User Group Meeting #4](https://github.com/GeoWerkstatt/interlis-check-service/wiki/Anwender:innen-Gruppe#user-group-meeting-4-18092024-1330-1430-uhr-per-teams) wurden die Resultate aus dem Teamcamp 2024 präsentiert. Darunter auch eine Differnzbildung für INTERLIS 2.4 Dateien. Hier werden Anwendungsmöglichkeiten und Problemstellungen zum Thema diskutiert.

**Comment by olivergrimm** [comment]
Protokoll zur heutigen Besprechung mit den Kantonen SH und SZ:

- Das Werkzeug soll in Richtung "ÖREB Richtigkeitsbescheinigung" unterstützen
- Beide Kunden sehen eine neue Rolle im Datenabgabe-Prozess "Beurteilung der Lieferung"
- Diese Rolle kann sowohl durch interne (Datenannahmestelle) wie auch externe (Datenlieferant) Stellen wahrgenommen werden.
- Das Feature könnte als "alleinstehendes" Werkzeug basierend auf einer Nachfolger-Vorgänger-Beziehung implementiert werden.
- Stabile Objektidentifikation wird als Voraussetzung für die Differenzerkennung verstanden.
- Es sind allfällige rechtliche Aspekte zu prüfen im Zusammenhang mit Öreb-Datensätzen
- Die Festlegung was relevante Differenzen sind, ist spezifisch vorzunehmen.

Nächste Schritte:
1. Workshop zur Thematik "Differenzdarstellung"
2. Implementierungsvorschlag durch geow, anschliessend Diskussion

**Reply by Philippluca** [reply]
Zusatz aus der Erinnerung:

- Für Darstellung der Differenzen scheint es mehre Optionen zu geben. Kombiniert in einer Karte, mit Slider (Vorher, Nacher), jeweils mit oder ohne Modell-speziefischer Symbologie.

**Comment by romefi** [comment]
Wir haben im Kanton Schaffhausen einen groben Ablauf zum Einsatz von Geopilot für Lieferung und Genehmigung von ÖREB-Datenlieferungen entworfen. Dort spielt die Differenzbildung eine zentrale Rolle. Der Ablauf ist als Entwurf zu verstehen, der nicht alle Aspekte beinhaltet und nicht alles zu Ende gedacht ist.

[Weiterentwicklung Geopilot - Ablauf Differenzbildung.pdf](https://github.com/user-attachments/files/18108174/Weiterentwicklung.Geopilot.-.Ablauf.Differenzbildung.pdf)

**Comment by Philippluca** [comment]
Hier noch ein Screenshot aus dem XTF Diff POC welcher im Rahmen des 2024 geowerkstatt-Hackathon entstanden ist:
![image](https://github.com/user-attachments/assets/a66869e3-3f28-4861-ae07-866325a95dbe)

**Comment by Philippluca** [comment]
Hier das Protokoll des Workshops: 
# Differenz Viewer Workshop 24.03.2025

Teilnehmer:
- Romedi Filli (SH)
- Delia Erb (SH)
- Vanessa Guidetta (SH)
- Kuno Epper (SZ)
- Michael Burkard (TG)
- Marc Schmid (BAZL)

## Einladungstext
**Einladung zum Workshop: Differenz-Viewer POC**

Liebe geopilot Kunden

wir freuen uns, euch zu einem Workshop zum Thema "Differenz-Viewer POC" einzuladen. Dieser Workshop wird in Zusammenarbeit mit Kt. Schaffhausen durchgeführt und richtet sich an alle, die am Projekt Differenz-Viewer POC mitwirken möchten.**Hintergrund**Der Prototyp eines Differenz-Viewer entstand während eines Hackathons/Teamcamp der geowerkstatt in 2024. Bei diesem Event wurde der Grundstein gelegt um INTERLIS 2.4 Dateien lesen und vergleichen zu können. Entstanden ist ein erster Prototyp, welcher die Differenzbildung für INTERLIS 2.4 Dateien demonstriert. Die Ergebnisse wurden im [User-Group Meeting #4](https://github.com/GeoWerkstatt/interlis-check-service/wiki/Anwender:innen-Gruppe) gezeigt. Nach der Demonstration der technischen Möglichkeiten und einem Austausch mit SH und SZ ([Protokoll](https://github.com/GeoWerkstatt/geopilot/discussions/318))) möchten wir nun die Anwendungsfälle für ein Differenz-Viewer Tool eruieren und die Entwicklung- und Finanzierungsmodalitäten festlegen.**Ziele des Workshops**

*   Festlegung der Fragen, welche das Tool beantworten soll. (Anwendungsfälle)
*   Festlegung der Rahmenbedingungen für die Umsetzung
*   Definition weiteres Vorgehen

**Agenda**

1.  Begrüßung und Einführung
2.  Sammlung und Konsolidierung der zu beantwortenden Fragen
3.  Diskussion mögliche Lösungsansätze / Darstellungsmodalitäten es gibt.
4.  Entwicklungsvorgehen mit Investoren klären & Abgrenzungen festlegen.
5.  Fakultativ gemeinsames Mittagessen (Ich bitte um eine An bzw. Abmeldung für das Mittagessen)

**Vorbereitung**

Bitte bereitet euch auf den Workshop vor, indem ihr folgende Punkte durchgeht:

1.  Lest das Protokoll zur letzten Besprechung mit den Kantonen SH und SZ. [(Protokoll)](https://github.com/GeoWerkstatt/geopilot/discussions/318)
2.  Schaut euch den groben Ablauf zum Einsatz von geopilot für Lieferung und Genehmigung von ÖREB-Datenlieferungen an. [(Ablauf Differenzbildung.pdf)](https://github.com/user-attachments/files/18108174/Weiterentwicklung.Geopilot.-.Ablauf.Differenzbildung.pdf)
3.  Überlegt euch welche Fragen in euren Prozessen durch ein Diff-Viewer beantworten müssen.

**Bekannte Abgrenzungen**

*   Der Differenz-Viewer soll als eigenständiges Tool funktionieren um 2 INTERLIS 2.4 XTF-Dateien vergleichen zu können.
*   Die Integration in geopilot wird angedacht, ist jedoch nicht Teil des Projektes Differenz-Viewer.

**Details**

*   Datum: 24. März 2025
*   Uhrzeit: 09:00 Uhr
*   Ort: geowerkstatt GmbH, Bleichemattstrasse 2, 5000 Aarau
*   Teilnehmer: Maximal 3 Teilnehmer pro Partei

Wir freuen uns auf eure Teilnahme und eine produktive Diskussion. 

Mit freundlichen Grüßen

Philipp Lüthi

# Erkenntnisse

## WOW-Erkenntnisse:

- Differenzen werden kontextbezogen beurteilt 
- Nicht alle Differenzen sind gleich relevant
- Differenz als eigenständiges Lieferobjekt ("Diffreport") / Bildung des Inkrements zur Historisierung

## Fachliche Erkenntnisse:
- Stabile OIDs reichen nicht aus / Rückfallebene mit konfigurierbaren stabilen identifizierenden Attributen. Keine weitere Diff-Berechnung nötig in Phase 1.
- Differenzen unterschieden sich in relevante und weniger relevante Differenzen. Die Klassifizierung ist fachspezifisch definiert (kontextbezogen)
    ○ Kontext: Person / Rolle, Fachmodell, Attributsbezug
- Darstellung bedingt: 2 fachspezifische Views (ALT/NEU) sowie Differenzdarstellung 
- Änderungsreport / Inkrement stellt ein Lieferobjekt dar (Bezug zu Auflageportal)
- Differenzen können sehr detailliert sein und sind entsprechend schwer darstellbar (pro Datenvergleich)
- Differenzanzeige muss vollständig alles anzeigen und die Möglichkeit zur Filterung bieten.
- Anwender können sehr GIS-nah sein und andererseits Stellen betreffen ohne technischen Bezug zu GIS
- muss INTERLIS 1 .. 2.4 unterstützen

## Priorisierung der Prozessfragen: 
1. Sind die inkrementellen Daten / der neue Zustand korrekt?
2. Wo hat sich was geändert?
3. Wie entsteht der neue Zustand?

# Weiteres Vorgehen
1. Grobe Lösungsspezifikation
    - Bildung Differenzen
    - Darstellung Datenbestand X auf Basis Fachmodell
    - Darstellung Differenzen (Karte + Liste)
    - UI Konzept
    - Schätzung 
    - Klärung Finanzierbarkeit 
2. Agile Umsetzung / MVP
3. Integration geopilot



![20250324_104651](https://github.com/user-attachments/assets/de533ea4-f532-47ca-8862-a9d0e4f60dcb)
![20250324_111620](https://github.com/user-attachments/assets/7b9abf21-6638-44c7-a3cb-1012660c7df0)
![20250324_112514](https://github.com/user-attachments/assets/cf818ce8-a81f-44ce-92b5-8d7784f19fb0)
![20250324_113245](https://github.com/user-attachments/assets/ef2f49f9-ec6f-40d9-8ac2-d4a51f8da559)
![20250324_135817](https://github.com/user-attachments/assets/a17e0970-c9ee-43f6-8b43-171f3544a613)
![20250324_135840](https://github.com/user-attachments/assets/d14bada5-3cc6-4dd4-abde-b7817a09fee8)
![20250324_135859](https://github.com/user-attachments/assets/4c96e915-3968-48f7-af6f-f5051f7d4419)

**Comment by Philippluca** [comment]
Statusupdate:
Der Workshop zum Differenz Viewer ist nun bereits einige Zeit her und ich dachte es ist wieder Zeit für ein Statusupdate.

Während dem Workshop sowie beim Einordnen dieser Erkenntnisse hat sich für uns gezeigt, dass die Thematik der Differenz Berechnung sowie der Bereitstellung und Darstellung eine gewisse Komplexität mitbringt. Im speziellen dann, wenn wir INTERLIS Dateien betrachten möchten, welche mit mehreren Klassen, Assoziationen und Structs unterschiedliche Arten von Objekten enthalten. Im Workshop haben wir als Ausganslage einen bekannten Identifikationsgrösse für Objekte angenommen. Wenn wir über die INTERLIS Repositories schauen, scheint dies keine sinnvolle Annahme zu sein. Wir verfolgen aktuell einen Lösungsansatz, welcher die Geometrien dynamisch abgleicht und daher nicht auf einen stabilen Identifikator zurückgreifen muss. Dies erlaubt ein Vergleich aller Modelle. 

Um die genannte Komplexität und mögliche Limitationen besser zu verstehen, erarbeiten wir gerade unabhängig einen POC, welcher uns wichtige Erkenntnisse für eine Schätzung geben soll. Wir werden diesen POC testweise in geopilot integrieren, um den Overhead von einem Zusatzprojekt zu umgehen.

Folgende Erkenntnisse und Ergebnisse gibt es bereits:

- Ili2pg ist Vorbedingung für eine Analyse direkt in der Datenbank. Das Einlesen kann je nach Menge und Komplexität der Daten mehrere Minuten in Anspruch nehmen.  (Soweit bekannt)
  - Zukünftig könnte hier eine Optimierung des Prozesses angestrebt werden.
- Mit PostGIS lässt sich ein Vergleich direkt auf der Datenbank sehr performant umsetzten.  
  - Annahme: Filter etc. können daher einfach eingebaut werden, um relevante und weniger relevante Änderungen zu unterscheiden. 
- In PostGIS lässt sich ein Vergleich so umsetzten, dass die Objekte über die Geometrie gematched werden. 
- Der Match auf die Geometrie funktioniert live, sofern ein Index vorhanden ist.
  - Der Match funktioniert auf für neue, gelöschte und leicht geänderte Geometrien.
  - Durch den Vergleich auf Geometrie-Ebene lassen sich zuverlässige Aussagen zu geänderten, gelöschten oder neu hinzugefügten Objekten machen.
  - Die Darstellung lässt sich in QGIS mit kleinen Umwegen umsetzten.

Hier demonstrativ eine View auf Daten der SH Grundnutzung:
![image](https://github.com/user-attachments/assets/1dc14c28-79fd-415a-860d-f0d3f2e8b657)

Ich melde mich wieder, sobald wir den POC mit statischen Daten bereit haben und für eine grobe Schätzung bereit sind. Danach werden wir mit den Projektparteien einen Projektplan erstellen.

**Comment by romefi** [comment]
Wenn ich die Diskussion von gestern aus meiner Sicht zusammenfasse, komme ich zu folgendem:

**Objektidentifikatoren**
Vorausgesetzt sind Objektidentifikatoren in den Datenmodellen. Ob diese als OID modelliert oder sonst ein Attribut Objektidentifikator ist, spielt im Moment keine Rolle. Aber es ist die Prämisse, dass es Objektidentifikatoren geben muss, damit eine vollständige Differenzbildung möglich ist.

**Output**
- Differenz alt <> neu: wird benöigtig, in der Differenz-View anzeigen zu können, was die Änderungen sind und so beurteilt werden kann, ob die Änderungen ok ist. Hier müssen alle Änderungen ausgegeben werden: Geometrie, Attribute, Relationen usw.
- Inkrement: Für den Differenz-Viewer ist das Inkrement nicht brauchbar, da es nur den neuen Zustand beschreibt und nicht die Differenz von alt zu neu. Das Inkrement kann jedoch für die Weiterverarbeitung (Datenupdate, Historisierung) genutzt werden.

Für das weitere Vorgehen bestehen aus meiner Sicht zwei Möglichkeiten:

**Variante A**
Die zwei Zeitstände werden in PostGIS importiert und über die DB-Funktionalitäten verglichen. Die Outputs werden in der DB bereitgestellt und könnten von dort aus weiter genutzt werden.
Der Vorteil ist, dass die Daten für eine fachspezifische Darstellung, die für den Differenzviewer gefordert wird, so oder so in die Datenbank importiert werden müssen.

**Variante B**
Ein Opensource-Tool wird entwickelt, das zwei XTF in verschiedenen Zeitständen vergleichen kann und die Differenz sowie Inkrement als Output liefert. Optimalerweise wieder als XTF, ggf. in einem eigenen INTERLIS-Modell.
Der Output wird zusammen mit Zustand alt und neu inPostGIS importiert und dem Differenzviewer bereitgestellt.
Der Vorteil wäre, dass dieses Tool auch ausserhalb der geopilot-Welt genutzt werden könnte für andere Anwendungszwecke.

Habe ich etwas vergessen oder falsch verstanden?

**Reply by agibur** [reply]
Die Variante B ist mit Sicherheit flexibler einsetzbar, zumal weniger von PostGIS abhängig als die Variante A. Die Variante B unterliegt dennoch einer gewissen Restriktion:

**Variante B**
Der Output des Opensource-Tools sollte von Vorteil direkt in PostGIS zwecks der Darstellung im Differenzviewer übernommen werden können, ohne dass dafür nochmals eine Differenzbildung der beiden XTF's durchzuführen ist. Für mich stellt sich bei diesem Ansatz nach wie vor die Frage, was hier der zweckmässigste Output des Tools ist, der ein möglichst breites Anwendungsspektrum abzudecken vermag.

Weitere offene Fragen:
- Welche andere Anwendungsfälle soll der Output unterstützen? Welche Besonderheiten gibt es dabei zu beachten?
- Wird PostGIS einzig für die Datenaufbereitung für den Differenzviewer eingesetzt und inwiefern wird die persistente Sicherung der Daten benötigt? Evtl. ausschliesslich im Zusammenhang mit der Richtigkeitsbestätigung für den ÖREB?
- Soll ili2pg als externe Komponente im Opensource-Tool eingebunden werden?

**Reply by Philippluca** [reply]
@romefi Aus meiner sicht passt deine Zusammenfassung sehr gut zu den Erkenntnissen aus unserem Meeting. Drei Anmerkungen:
- Open-Source können beide Tools entwickelt werden. 
- Variante A beschreibt ein Tool, welches die Abfragen für die Differenzen automatisch generiert basierend auf vorherrschendem Schema.  
- Variante A erlaubt es direkt auf der Datenbank auf Geometrievergleiche zu erstellen. 

Welche Variante bzw. welches Vorgehen ziehst du nun vor? Für Variante B ist impliziert, dass der Anwendungsfall nicht primär durch einen  Diff-Viewer gegeben ist. => Kennen wir also die Anforderungen oder ist dies eher ein Wunsch?

**Reply by romefi** [reply]
Danke @Philippluca und @agibur für die Rückmledungen.

Open-Source sind beide, richtig. Variante A braucht aber zwingend PostGIS dafür, Variante B würde ohne PostGIS laufen und wäre daher unabhängiger.

Der Output müsste durch die Anforderungen vom Differenzviewer gegeben sein. Wobei es dann eigentlich egal ist, ob der Output per Variante A direkt auf der DB generiert wird oder über Variante B mit einem Standalone-Tool, das die Differenz bildet und den Output generiert. Könnte wiederum ein XTF nach vorgegebenem Datenmodell sein, das den Output klar definiert (konzeptionell anaolg dem errolog.xtf von ilivalidator). Dieser Output könnte dann wieder in PostGIS importiert werden und für den Differenzviewer genutzt werden.

Für mich ist der Anwendungsfall von einem Diff-Viewer bei beiden Varianten gegeben und ist weiterhin das Ziel. Einziger Unterschied ist der Weg dorthin. Direkt auf der DB mit Variante A oder mit Variante B über ein Standalone-Tool, das die Differenz bildet und somit auch für andere Anwendungsfälle, in denen die Differenz verlangt wird, eingesetzt werden könnte.

**Reply by KunoEpper** [reply]
Für mich hatte der Differenz-Viewer den Zweck, den Unterschied zwischen zwei xtf-Dateien zu erkennen. Als Output kann ich mir zwei Arten vorstellen:
- Output 1: **Eine XTF-Datei**, welche die Änderungsarten pro Objekt ausweist (Inkrement)
- Output 2: **Eine XTF-Datei pro Änderungsart** (z.B. "neu.xtf", welche alle neuen Objekte enthält; siehe Abbildung unten)
![Output-Arten](https://github.com/user-attachments/assets/1f414b25-311f-4126-a8cc-6736bc59c7a1)

Dies wäre ähnlich der **Variante B** von oben.
Anwendungsfall:
- Vergleich von XTF-Dateien: Worin unterscheiden sich die Objekte in der neu gelieferten Datei im Vergleich zur Vorgängerlieferung?
- Deltaberechnung/Archivierung/Historisierung: Bilde mir ein Inkrement. Dieses kann für die Versionierung/Archivierung/Historisierung verwendet werden (geringere Dateigrösse, falls man die "unveränderten Objekte" weglässt).
![Inkremente](https://github.com/user-attachments/assets/53d4dae2-fecc-4a9b-b8bb-d90eb78e40fa)

Aus V0.xtf und V1.xtf wird das Inkrement Ink1.xtf berechnet (usw.). V2.xtf ergibt sich aufgrund V0.xtf plus den Änderungen in Ink1.xtf (usw.).

Wie erwähnt: Das war meine ursprüngliche Auffassung des Differenz-Viewers. Der Wunsch, diesen im ÖREB-Prozess für die Darstellung des Änderungsplans zu nutzen, hätte zweifelsohne einen grossen Mehrwert.

Bei Nutzungsplanänderungen wird ein Änderungsplan erstellt. Soweit ich Beispiele aus dem Kanton Schwyz interpretiere, weist dieser Änderungsplan nur die Geometrieteile auf, welche sich effektiv ändern. Dies im Unterschied zur oben beschriebenen Lösung, bei der die ganze Geometrie als "geändert" angezeigt würde.

In der nachfolgenden Abbildung habe ich die verschiedenen Änderungen zusammengestellt. Die in der Abbildung als "Inkrement" aufgeführten Flächen entsprechen in etwa der Variante B. Die als "Differenz" aufgeführten Flächen der Variante A.
![Änderungen](https://github.com/user-attachments/assets/d12361ab-db32-42c8-a223-f5120351133a)

Die Variante A ermöglicht die Bestimmung der roten und grünen Teilflächen über PostGIS-Funktionen und ist Voraussetzung für die Erstellung des Änderungsplans. Es gibt jedoch Fälle, wo entschieden werden muss, welche Änderungsart angezeigt wird (das Hinzufügen des neuen, grünen Flächenteils oder der Verlust des alten, roten Flächenteils; vgl. "Beispiel mit Gebietseinteilungen").

**Fazit**
- Ich möchte an der Funktion festhalten, dass unser Tool die Unterschiede zwischen zwei xtf-Dateien bestimmen kann (Inkrement).
- Ich bin der Meinung, dass die Funktion zur Berechnung der "Differenz" nicht im Tool realisiert werden soll, da sie auf PostGIS vorhanden ist und dessen Nutzung für mich zweckmässig und kostengünstiger erscheint.
- Ich könnte mir vorstellen, dass für die Erstellung eines automatisierten Änderungsplanes noch die eine oder andere Hürde zu überwinden ist. Um Überraschungen zu verhindern, sollte die Gruppe ein Regelwerk bestimmen, welche Änderung im Diff-Viewer wie angezeigt werden soll.

**Reply by romefi** [reply]
@KunoEpper nur um sicher zu gehen, ob ich dich richtig verstanden habe. Dann würdest du die Differenzbildung gemäss Variante A in PostGIS favorisieren gegenüber Variante B, die Differenzbildung in einem eigenständigen Tool zu berechnen?

Oder Mischform, d.h. Differenzbildung in PostGIS, jedoch mit XTF-Outputs?

**Reply by KunoEpper** [reply]
@romefi 
Vielleicht zuerst zur Begrifflichkeit:
- Als "Differenz" im Kontext dieser Diskussion verstehe ich "die Berechnung der **Teil**flächen", also das Bestimmen der roten und grünen Flächen in der Abbildung oben.
- Für mich ist eine "Differenz" aber auch als "Unterschied" zu verstehen. Wenn ich ein Tool habe, welches mir den Unterschied zwischen zwei Dateien bestimmt, habe ich auch eine Differenz. Im aktuellen Kontext nutze ich aber dafür den bereits eingeführten Begriff "Inkrement".

Grundsätzlich möchte ich ein Standalone-Tool. Da es offenbar zu komplex würde oder so nicht im Fokus war oder damit noch andere Fragen zu klären sind, würde ich die Differenzberechnung aus Kostengründen auf PostGIS favorisieren. Die Kostenunteschiede würden mich aber interessieren. Und wer weiss: Vielleicht gibt es eine Library, welche die Funktion auch auf Stufe Tool ermöglichen würde. Dann würde ich schwenken und alles im Tool realiseren wollen.

**Comment by olivergrimm** [comment]
Wir erarbeiten nun eine Lösungsbeschreibung / Offerte für ein standalone Werkzeug zur Detektierung der Differenzen. Dieses soll bewusst eine Phase 1 darstellen und den Teil _Viewing_ sowie _Integration in geopilot_ bewusst ausschliessen.
Output dieses Werkzeugs ist eine Datei, welche die detektierten Differenzen in einem konsolidierten Modell und im Format JSON aufführt.
Rahmenbedingung dafür ist, dass die OIDs stabil und damit vergleichbar sind. Die unterstützten INTERLIS-Versionen sind 2.3 und 2.4.
